### ISSUE TITLE
Continue training or train from begining, with modified/increased training/val images ...

### ISSUE THREAD

**Issue by VanitarNordic** [AlexeyAB/darknet#20]
Hello,

I've heard if we gather those images which a model finds **false positive** objects inside, then annotate and add them inside existing images dataset, then we can increase the accuracy of the model **dramatically**.

in this case, assume that we have trained the model for 2000 iterations and we have a weight file of this. Then we add these new images into the dataset. 

Should we start training from the beginning (from 1) or we can continue training by 2001?

**Comment by AlexeyAB** [comment]
@VanitarNordic 

**It is better to train from the beginning.** But not necessarily.

* If you will continue to train from 2001 iteration only with dataset of new 500 images, then the network will gradually learn new images, **but gradually forget old images.**

* If you have 2000 images and has trained 2000 iterations, then find new images with false positive or false negative detections, for example 500 images, then you can add it to first dataset and continue training from 2000 iterations with this 2500 images, **but too many iterations can lead to overfitting,** i.e. you will can to detect only images from the dataset: https://en.wikipedia.org/wiki/Overfitting

Example of overfitting:
![overfitting](https://cloud.githubusercontent.com/assets/4096485/22627161/7a02f4dc-ebce-11e6-91cc-362213c2b7c5.png)


You should necessarily to train from the beginning only if your dataset is wrong, or if you change your Network `.cfg`-file, for example, if you change number of objects (classes).

**Comment by VanitarNordic** [comment]
Yess, I am agree with you, from the beginning is better, at least to avoid over-fitting as you mentioned correctly. Brilliant. Please check the YOLO_Mark also, I have posted a small issue there.

**Comment by VanitarNordic** [comment]
is there anyway to verify the **validation set** errors during training? I mean which parameter in the darknet Yolo-2 training belongs to that?
`
avg error` seems to be for Training set.

**Comment by AlexeyAB** [comment]
@VanitarNordic 

No. To validate, you should stop training, get one of weights, and do the validation:
* put path-names of validation images to file `data\voc.2007.test` (or simply copy `data\train.txt` to `data\voc.2007.test`)
* `darknet.exe detector recall data/obj.data yolo-obj.cfg yolo-obj_9000.weights -i 0`

You will get such result, where last line is main:

>  7580  7606  7683       RPs/Img: 68.21  IOU: 77.86%     Recall:99.00%
>  7581  7607  7684       RPs/Img: 68.21  IOU: 77.86%     Recall:99.00%
>  7582  7608  7685       RPs/Img: 68.21  IOU: 77.86%     Recall:99.00%
>  7583  7609  7686       RPs/Img: 68.22  IOU: 77.86%     Recall:99.00%
>  7584  7610  7687       RPs/Img: 68.22  IOU: 77.86%     Recall:99.00%
>  7585  7611  7688       RPs/Img: 68.22  IOU: 77.86%     Recall:99.00%
>  **7586  7612  7689       RPs/Img: 68.23  IOU: 77.86%     Recall:99.00%**
> 
> D:\Darknet2\darknet\build\darknet\x64>pause

**Comment by VanitarNordic** [comment]
it worked, in my case:

![2017-03-01_14-03-29](https://cloud.githubusercontent.com/assets/18719591/23456246/e72f972e-fe87-11e6-9dc4-01dac35776fd.jpg)

**Comment by VanitarNordic** [comment]
What are **RPs/Img** and **IOU** values and we should try to make them smaller or bigger?

**Comment by AlexeyAB** [comment]
RPs/Img, IOU and Recall calculated here: https://github.com/AlexeyAB/darknet/blob/2a9a5229c87c5e05c87d9d792c62cf020b3f1981/src/detector.c#L432

* **RPs/Img** - average found objects per image (nothing say about accuracy)
* **IOU** - the bigger, the better (says about accuracy)
* **Recall** - the bigger, the better (says about accuracy)

![precision_recall_iou](https://cloud.githubusercontent.com/assets/4096485/23457648/dfc08a64-fe89-11e6-91b7-17ad333d156d.jpg)

**Comment by VanitarNordic** [comment]
How many training and validation images are typically needed to make a good detector? 

because in my case, I have made a one class apple detector model with a small dataset, although calculations shows very good avg error and IOU, but the model still detects many false-positives.

**Comment by AlexeyAB** [comment]
@VanitarNordic 

* training images required about 2000 per class (object)
* validation images required about 500 per class (object)

I added some manual **When should I stop training**: https://github.com/AlexeyAB/darknet#when-should-i-stop-training

You can put here some of image from training data-set and false-positives results, as example, may be I can say something about it.

**Comment by VanitarNordic** [comment]
this is an example of a false detection. it has detected the back of a monitor as apple. anyway dataset is very small, it might be natural (130/25 training/validation images). I make these for experiment

![predictions](https://cloud.githubusercontent.com/assets/18719591/23460664/b7a40d1e-fe9b-11e6-99c6-25b395151d8f.jpg)

**Comment by AlexeyAB** [comment]
Try to add all images with false-positive to training dataset.

**Comment by VanitarNordic** [comment]
> Try to add all images with false-positive to training dataset.

Yes, that's correct. I am agree.

Also, one question in side, Does amount of computer RAM (not the GPU memory) affects the training? 

When I train, the computer becomes unstable and slow (once dropped of the training in the middle of the job around few hundreds of iterations, I did not see the error and reason of that).

I have 8G DDR4-2400 RAM installed, I'm curious if increasing it to 16G would solve the case, specially in future when I want to train bigger datasets.

the image shows consumption of resources, considering Windows 10 itself consumes around 2-3G of RAM, should I increase it?

![2017-03-01_17-33-43](https://cloud.githubusercontent.com/assets/18719591/23463025/6134efac-fea5-11e6-8489-055f67379e18.jpg)

**Comment by VanitarNordic** [comment]
another interesting point is when I process a single image, it says it is predicted in 0.024 second. It means that I should get around 41 FPS in real time but in practice I get variations between 8 to 20FPS, why?!

**Comment by VanitarNordic** [comment]
The number is correct when I process a video file, close to calculations, 38 to 40FPS

**Comment by AlexeyAB** [comment]
Network camera can send only that numbers of frames, which you set on it :)

**Comment by VanitarNordic** [comment]
hahaha, yes it solved but it varies too much. for a video file it is more stable. it varies from 5 to 39FPS.

The only two issues remained here at the moment one is detection of small objects which I have problem with and also making some GUI application based on your repository.

**Comment by AlexeyAB** [comment]
About small object, there is solution - **1.2. If you want to detect small object on images with 832x832 or 1088x108:** https://github.com/AlexeyAB/darknet/issues/12#issuecomment-283340030

**Comment by VanitarNordic** [comment]
That was a weird error which I told you above. Now I catch it and got a picture. The reason is unknown.

![2017-03-02_12-19-33](https://cloud.githubusercontent.com/assets/18719591/23499892/31e90f78-ff43-11e6-849b-c3b592ecebce.jpg)

**Comment by AlexeyAB** [comment]
Did you do on your computer for something else at the same time?
I think something else took the GPU-RAM or the GPU-RAM.

**Comment by VanitarNordic** [comment]
no nothing, I close all programs when I decide to train. besides as you see in the picture in some comments above, only a half of GPU was occupied. and computer becomes slow and sometimes mouse pointer has difficulties in movement but it does not freeze. This phenomenon intensifies if I use high res training images. It does not affect GPU memory consumption but it affects system performance.
